Java/parkservices/src/main/java/com/amazon/randomcutforest/parkservices/GlobalLocalAnomalyDetector.java

```diff
@@ -0,0 +1,410 @@
+                " too few or too many clusters are not " + "meaningful to this algorithm");
+        maxAllowed = value;
+    }
+
+    /**
+     * The following provides a single invocation for scoring and updating.
+     * Semantics of the recency biased sampling (sequentiality in decision making)
+     * and efficient automatic reclustering demand that scoring and updating be
+     * simultaneous. While scoring is provided as a separate function to let future
+     * preditor-corrector methods reuse this code, it is strongly recommneded that
+     * only the process() function be invoked.
+     * 
+     * @param object            current object being considered
+     * @param weight            weight of the object (for clustering purposes as
+     *                          well as recency biased sampling)
+     * @param localDistance     a local distance metric that determines the order in
+     *                          which different clusters are considered; can be
+     *                          null, in which case the global distance would be
+     *                          used
+     * @param considerOcclusion consider occlusion by smaller dense clusters, when
+     *                          adjacent to larger and more spread out clusters
+     * @return a generic descriptor with score, anomaly grade, nearest
+     *         representative and threshold (anomaly grade greater than zero is
+     *         likely anomalous; anomaly grade can be -ve to allow down stream
+     *         correction using semi-supervision or other means)
+     */
+    public GenericAnomalyDescriptor<P> process(P object, float weight, BiFunction<P, P, Double> localDistance,
+            boolean considerOcclusion) {
+        checkArgument(weight >= 0, "weight cannot be negative");
+        // recompute clusters first; this enables easier merges and deserialization
+        if (sequenceNumber > lastCluster + doNotreclusterWithin) {
+            checkArgument(globalDistance != null, "set global distance function");
+            double currentMean = thresholder.getPrimaryDeviation().getMean();
+            if (abs(currentMean - lastMean) > 0.1 || currentMean > 1.7
+                    || sequenceNumber > lastCluster + 20 * doNotreclusterWithin) {
+                lastCluster = sequenceNumber;
+                lastMean = currentMean;
+                clusters = getClusters(maxAllowed, 4 * maxAllowed, 1, numberOfRepresentatives, shrinkage,
+                        globalDistance, null);
+            }
+        }
+        Weighted<P> result = score(object, localDistance, considerOcclusion);
+
+        double threshold = thresholder.threshold();
+        double grade = 0;
+        if (result.weight > 0) {
+            grade = thresholder.getAnomalyGrade(result.weight, false);
+        }
+
+        // note average score would be 1
+        thresholder.update(result.weight, min(result.weight, thresholder.getZFactor()));
```

Comment:
It just clips to score so that very large scores do not get to change the threshold. Currently there is no good analogue of score difference (because different points are not connected in time).